### PR TITLE
CCUBE-1218

### PR DESCRIPTION
**Pull request by ShIrInRuTh**
**Changes**
1. Merged the changes from CCUBE-1213 into this branch so as to use the mock context for the tests to be created for CCUBE-1218
2. Added new tests for the side-panel, side-panel-header & toolbar

**Additional information**

-   You may refer to this [CCUBE-1218](https://jira.ship.gov.sg/browse/CCUBE-1218)
-  ** Some of the tests in the side panel, side-header-panel test files may not work as they rely on the new chnages made in CCUBE-1202 and other branches that have yet to be merged to pass, so will need to run the tests again to check if they work oce the relevant changes have been merged to main**

### CODE REVIEW

File: tests/components/side-panel/side-panel-header.spec.tsx [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+            renderComponent({
+                builderContext: {
+                    ...mockBuilderState,
+                    focusedElement: {
+                        element: mockElement,
+                    },
+                },
+            });
+            expect(getHeaderLabel().textContent).toEqual("Edit details");
+        });
```
Comment (weili-govtech) [inline tests/components/side-panel/side-panel-header.spec.tsx]: test is failing

File: tests/components/side-panel/side-panel-header.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    });
+});
```
Comment (weili-govtech) [inline tests/components/side-panel/side-panel-header.spec.tsx:55]: should also test for rendering of `Add pages`

File: tests/components/side-panel/toolbar.spec.tsx [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        jest.restoreAllMocks();
+        jest.resetAllMocks();
+    });
+
+    describe("viewToggleMode()", () => {
+        it("should display the relevant mode when toggleView function is called", () => {
+            renderComponent();
+            fireEvent.click(getIconButton());
+            expect(mockToggleMode).toHaveBeenCalled();
+        });
```
Comment (weili-govtech) [inline tests/components/side-panel/toolbar.spec.tsx]: can remove as it is not worthwhile to test

File: tests/components/side-panel/toolbar.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            renderComponent({
+                builderContext: {
+                    ...mockBuilderState,
+                    showSidePanel: false,
+                },
+            });
+            fireEvent.click(getIconButton());
+            expect(mockTogglePanel).toHaveBeenCalledWith(true);
+            expect(mockToggleMode).toHaveBeenCalled();
+        });
```
Comment (weili-govtech) [inline tests/components/side-panel/toolbar.spec.tsx]: - description should be `should expand toggle panel if it is collapsed on clicking the mode button`
- there should be another test for `should collapse toggle panel if it is expanded on clicking the mode button`

File: tests/components/side-panel/toolbar.spec.tsx [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+import { Toolbar } from "src/components/side-panel/toolbar";
+import { TestHelper, mockBuilderState } from "src/util/test-helper";
+
+describe("toolbar", () => {
+    afterEach(() => {
+        jest.restoreAllMocks();
+        jest.resetAllMocks();
+    });
+
+    describe("viewToggleMode()", () => {
```
Comment (weili-govtech) [inline tests/components/side-panel/toolbar.spec.tsx]: no need to wrap in this test suite

File: tests/components/side-panel/side-panel.spec.tsx [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        jest.resetAllMocks();
+    });
+
+    describe("rendering of the toolbar", () => {
+        it("should display the toolbar when there is no focused element", () => {
+            renderComponent();
+            expect(getToolbar()).toBeInTheDocument();
+        });
+
+        it("should not display the toolbar when there is a focused element", () => {
```
Comment (weili-govtech) [inline tests/components/side-panel/side-panel.spec.tsx:20]: test is failing

File: tests/components/side-panel/side-panel.spec.tsx [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+// =============================================================================
+// HELPER FUNCTIONS
+// =============================================================================
+
+const renderComponent = (overrideOptions?: TestHelper.RenderOptions) => {
+    return render(TestHelper.withProviders(overrideOptions, <SidePanel />));
+};
+
+const getToolbar = (useQuery = true) =>
+    useQuery ? screen.getByTestId("toolbar") : screen.queryByTestId("toolbar");
```
Comment (weili-govtech) [inline tests/components/side-panel/side-panel.spec.tsx]: wrong if else logic, if useQuery is true, it should not be using `getByTestId`

File: tests/components/side-panel/side-panel.spec.tsx [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+        it("should display the default content of the side panel if there is no mode given", () => {
+            renderComponent({
+                builderContext: {
+                    ...mockBuilderState,
+                    mode: EBuilderMode.EDIT_ELEMENT,
+                },
+            });
+            expect(getDefaultContent()).toBeInTheDocument();
+        });
```
Comment (weili-govtech) [inline tests/components/side-panel/side-panel.spec.tsx]: not needed, the default content panel is just a placeholder, so there is no point testing for its presence

File: tests/components/element-card/element-card.spec.tsx [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        jest.restoreAllMocks();
+        jest.resetAllMocks();
+    });
+
+    describe("element-card", () => {
+        it("should be in a focused state and show the edit element panel when clicked on", () => {
+            renderComponent({ element: mockElement, onClick: mockOnClick }, {});
+            fireEvent.click(getElementCard() as HTMLElement);
+            expect(mockOnClick).toHaveBeenCalled();
+        });
```
Comment (weili-govtech) [inline tests/components/element-card/element-card.spec.tsx]: description should be `should call onClick prop when clicked on`

File: tests/components/element-card/element-card.spec.tsx [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+        it("should not show the duplicate and delete buttons when in normal state", () => {
```
Comment (weili-govtech) [inline tests/components/element-card/element-card.spec.tsx]: `should not show the duplicate and delete buttons by default`

**Review by weili-govtech — CHANGES_REQUESTED** [review]

File: tests/components/side-panel/toolbar.spec.tsx [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    afterEach(() => {
+        jest.restoreAllMocks();
+        jest.resetAllMocks();
+    });
+
+    it("should collapse toggle panel if it is expanded on clicking the mode button", () => {
+        renderComponent();
+        fireEvent.click(getIconButton());
+        expect(mockToggleMode).toHaveBeenCalled();
+    });
```
Comment (weili-govtech) [inline tests/components/side-panel/toolbar.spec.tsx]: remove this test as we are not collapsing the panel on clicking any of the mode buttons

File: tests/components/side-panel/side-panel-header.spec.tsx [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+                builderContext: {
+                    focusedElement: {
+                        element: mockElement,
+                    },
+                },
+            });
+            expect(getHeaderLabel().textContent).toEqual("Edit details");
+        });
+    });
+});
```
Comment (weili-govtech) [inline tests/components/side-panel/side-panel-header.spec.tsx:55]: add test for the conditional rendering of the cross icon

File: tests/components/side-panel/side-panel-header.spec.tsx [hunk 13]
```diff
@@ -0,0 +1,63 @@
+import { render, screen } from "@testing-library/react";
+import "jest-canvas-mock";
+import { SidePanelHeader } from "src/components/side-panel/side-panel-header";
+import { EBuilderMode, EElementType } from "src/context-providers";
+import { ELEMENT_BUTTON_LABELS } from "src/data/elements-data";
+import { TestHelper } from "src/util/test-helper";
+
+describe("side-panel-header", () => {
```
Comment (weili-govtech) [inline tests/components/side-panel/side-panel-header.spec.tsx]: apologies, I just notice the incorrect naming, may I request to use the component name as the description of the main test suite, so this should be `SidePanelHeader`

do apply this to the rest of the test suites including the recently approved ones in `element-card`

**Review by weili-govtech — CHANGES_REQUESTED** [review]

File: tests/components/element-card/element-card.spec.tsx
Comment (weili-govtech) [inline tests/components/element-card/element-card.spec.tsx:1]: remove unused `mockBuilderState` import and actually it should not be exported at all

File: tests/components/side-panel/side-panel-header.spec.tsx (hunk 12 above)
Comment (weili-govtech) [reply tests/components/side-panel/side-panel-header.spec.tsx:55]: also test for it to be conditionally hidden

Comment (ShIrInRuTh) [reply tests/components/side-panel/side-panel-header.spec.tsx:55]: image

I created the test alrdy, its been pushed and it passes as well :)

Comment (weili-govtech) [reply tests/components/side-panel/side-panel-header.spec.tsx:55]: thanks while we are testing for its existence, we also need to test for scenario where it is not rendered, hence we need another test for it to be conditionally hiddden

**Review by weili-govtech — CHANGES_REQUESTED** [review]

File: tests/components/element-card/element-card.spec.tsx
Comment (weili-govtech) [reply tests/components/element-card/element-card.spec.tsx:1]: this is not updated?

**Review by weili-govtech — APPROVED** [review]

**Pull request merged**